src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs

```diff
@@ -47,14 +47,17 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
             // first we can make a small optimization:
             // If left is a constant then we already know whether it is null or not. If it is null then we 
             // can simply generate "right". If it is not null then we can simply generate
-            // MakeConversion(left).
+            // MakeConversion(left). This does not hold when the left is an unconstrained type parameter: at runtime,
+            // it can be either left or right depending on the runtime type of T
 
-            if (rewrittenLeft.IsDefaultValue())
+            var isUnconstrainedTypeParameter = rewrittenLeft.Type?.IsReferenceType == false && rewrittenLeft.Type?.IsValueType == false;
```

Comment:
Can `rewrittenLeft.Type` be `null` here? If not, consider replacing `?.` with `.`. If the type can be `null`, we only need to check once. (It may be clearer to check `rewrittenLeft.Type != null && …`.)

Reply:
Yeah, for the null literal. Will do.

---
In reply to: [213787027](https://github.com/dotnet/roslyn/pull/29577#discussion_r213787027) [](ancestors = 213787027)